Plurimath documentation blog (#19)
plurimath/plurimath.github.io: _posts/2023-08-22-using-plurimath.adoc

To get started, add this line into your applications gemfile.
add this line into your applications gemfile

```ruby
  gem "plurimath"
```
then run following command
```ruby
  bundle install
```
or run following command
```ruby
  gem install plurimath
```

Once it's incorporated, you can seamlessly convert mathematical expressions between various markup languages.

=== 1.1 Understanding Plurimath

*Plurimath* provides a range of conversion methods, allowing you to specify the input type and obtain the desired output format.
With just a few lines of code, you can harness the power of *Plurimath* to ensure consistency and compatibility in mathematical presentations across different platforms and applications.

Now let's discuss *how to use Plurimath?*.

```ruby
  formula = Plurimath::Math.parse(string, :type)
```

You will have to provide input text replacing `string`, and the `type` field should specify the type of the provided text, choosing from one of the supported languages.

The above example will generate a Formula object, something like this

```ruby
  Plurimath::Math::Formula.new([
    ... # Others objects
  ])
```

> Now next step would be conversion(since you parsed your string into *Plurimath Formula*) to your desired output language. For conversion all you have to do is call method name containing language's name with leading `to_`.

Reviewer: ```suggestion
Now the next step would be conversion (since you have parsed your string into a *Plurimath Formula*) to your desired output format. For conversion all you have to do is call the method name containing the format name with a leading `to_`.
```

```ruby
  formula.to_<language-name>
```
or you can do it directly,
```ruby
  Plurimath::Math.parse(string, :type).to_<language-name>
```

NOTE: Some methods support explicit optional argument and some don't.

> Below sections will be explaining each supported languages conversion and parsing with an example

Reviewer: ```suggestion
The sections below will explain each supported languages conversion and parsing with an example
```

=== 1.2 Use of Plurimath, Explained

> Below sections will help understand using *Plurimath* with realtime example.

Reviewer: ```suggestion
The sections below will help you understand using *Plurimath* with a real worked example.
```

We have an *AsciiMath* equation `sum_(i=1)^n i^3`, we will be using this equation in all the sections.

==== 1.2.1 AsciiMath to Formula

```ruby
  input_string = 'sum_(i=1)^n i^3'
  formula = Plurimath::Math.parse(input_string, :asciimath)
```

> Parsing the equation will generate following formula.

Reviewer: ```suggestion
Parsing the equation will generate the following formula.
```

```ruby
  Plurimath::Math::Formula.new([
    Plurimath::Math::Function::Sum.new(
      Plurimath::Math::Formula.new([
        Plurimath::Math::Symbol.new("i"),
        Plurimath::Math::Symbol.new("="),
        Plurimath::Math::Number.new("1")
      ]),
      Plurimath::Math::Symbol.new("n"),
      Plurimath::Math::Function::Power.new(
        Plurimath::Math::Symbol.new("i"),
        Plurimath::Math::Number.new("3")
      )
    )
  ])
```

==== 1.2.2 Formula to AsciiMath

> Following example follows the same steps we have before for *AsciiMath* as well.

Reviewer: ```suggestion
The following example follows the same steps we have before for *AsciiMath* as well.
```

```ruby
  input_string = <<~MATHML
    <math xmlns="http://www.w3.org/1998/Math/MathML" display="block">
      <mstyle displaystyle="true">
        <mrow>
          <munderover>
            <mo>&#x2211;</mo>
            <mrow>
              <mi>i</mi>
              <mo>=</mo>
              <mn>1</mn>
            </mrow>
            <mi>n</mi>
          </munderover>
          <msup>
            <mi>i</mi>
            <mn>3</mn>
          </msup>
        </mrow>
      </mstyle>
    </math>
  MATHML
  formula = Plurimath::Math.parse(input_string, :mathml)
  formula.to_asciimath # >> "sum_(i=1)^n i^3"
```

> NOTE: Heredoc is used above for better readability, but you're also welcome to use quotes to create strings if you prefer.

Reviewer: is this really needed? We assume readers know Ruby, which means they know that heredocs are strings.

*AsciiMath* grammar is different than other languages.

> The list of symbols and presentations available in *AsciiMath* is limited than other languages.

Reviewer: ```suggestion
The list of symbols and presentations available in *AsciiMath* is limited compared to other languages, and there is variability between different software in what symbols they support. Metanorma has drawn on Asciidoctor Asciimath in its implementation.
```

> You can see the list of symbols, classes and presentations available in *AsciiMath* at https://github.com/plurimath/plurimath/blob/main/MathML-Supported-Data.adoc[Supported *AsciiMath* Data].

Reviewer: ```suggestion
You can see the list of symbols, classes and presentations available in Metanorma's version of *AsciiMath* at https://github.com/plurimath/plurimath/blob/main/MathML-Supported-Data.adoc[Supported *AsciiMath* Data].
```

==== 1.2.3 Formula to MathML

A Formula object can be converted to *MathML* using the `to_mathml` method.

```ruby
formula.to_mathml
```
This will generate following output of *MathML*

```ruby
  <math xmlns="http://www.w3.org/1998/Math/MathML" display="block">
    <mstyle displaystyle="true">
      <mrow>
        <munderover>
          <mo>&#x2211;</mo>
          <mrow>
            <mi>i</mi>
            <mo>=</mo>
            <mn>1</mn>
          </mrow>
          <mi>n</mi>
        </munderover>
        <msup>
          <mi>i</mi>
          <mn>3</mn>
        </msup>
      </mrow>
    </mstyle>
  </math>
```

> In *MathML* we have *displaystyle* boolean attribute making the equation `normal` and `compact`.

Reviewer: ```suggestion
In *MathML* we have the *displaystyle* boolean attribute, making the equation `normal` or `compact`.
```

By default we use `display_style` true, but you can explicitly change it by passing `display_style:` argument to `to_mathml` method, depending on what output you desire.
See examples below.

```ruby
  formula.to_mathml(display_style: false)
  formula.to_mathml(display_style: true)
```
NOTE: *display_style* will affect the whole Formula equation and not any specific tag or part of the equation.

> NOTE: You can pass *display_style* attribute from string but it has to be top-level `mstyle` tag as shown in the example above.

Reviewer: ```suggestion
NOTE: You can pass *display_style* attribute within the string argument, but it has to be set at the top-level `mstyle` tag, as shown in the example above.
```

> You can see the list of https://github.com/plurimath/plurimath/blob/main/MathML-Supported-Data.adoc[*MathML* Supported Data] like symbols, tags and functions.

Reviewer: ```suggestion
You can see the list of https://github.com/plurimath/plurimath/blob/main/MathML-Supported-Data.adoc[*MathML* Supported Data] such as symbols, tags and functions.
```

==== 1.2.4 Latex to AsciiMath

Example of parsing and conversion for *LaTeX* below.

```ruby
  input_string = "\\prod_{\\theta}^{100}"
  formula = Plurimath::Math.parse(input_string, :latex)
  formula.to_asciimath # >> "sum_(i=1)^n i^3"
```

*LaTeX* encompasses an extensive array of symbols and functions, catering to various mathematical expressions.

> We support a comprehensive list of these functions and symbols, which you can be seen https://github.com/plurimath/plurimath/blob/main/Latex-Supported-Data.adoc[Supported *LaTeX* Data].

Reviewer: ```suggestion
We support a comprehensive list of these functions and symbols, which you can see at https://github.com/plurimath/plurimath/blob/main/Latex-Supported-Data.adoc[Supported *LaTeX* Data].
```

==== 1.2.5 OMML to AsciiMath

> OMML example processing below will also follow same steps.

Reviewer: ```suggestion
Processing the OMML example below will also follow the same steps.
```

          <m:sub>
            <m:r>
              <m:t>i</m:t>
            </m:r>
            <m:r>
              <m:t>=</m:t>
            </m:r>
            <m:r>
              <m:t>1</m:t>
            </m:r>
          </m:sub>
          <m:sup>
            <m:r>
              <m:t>n</m:t>
            </m:r>
          </m:sup>
          <m:e>
            <m:sSup>
              <m:sSupPr>
                <m:ctrlPr>
                  <w:rPr>
                    <w:rFonts w:ascii="Cambria Math" w:hAnsi="Cambria Math"/>
                    <w:i/>
                  </w:rPr>
                </m:ctrlPr>
              </m:sSupPr>
              <m:e>
                <m:r>
                  <m:t>i</m:t>
                </m:r>
              </m:e>
              <m:sup>
                <m:r>
                  <m:t>3</m:t>
                </m:r>
              </m:sup>
            </m:sSup>
          </m:e>
        </m:nary>
      </m:oMath>
    </m:oMathPara>
  OMML
  formula = Plurimath::Math.parse(input_string, :omml)
  formula.to_asciimath # >> "sum_(\"i\" = 1)^(\"n\") \"i\"^(3)"
```

We also support *OMML*, allowing you to work with mathematical content efficiently in *Office* documents.

> NOTE: *MathML* string supports *displaystyle* but *OMML* don't have any attribute or tag to understand it, so you will have to pass it explicitly to `to_omml(display_style: (true or false))`

Reviewer: ```suggestion
NOTE: *MathML* string supports *displaystyle* but *OMML* don't have any one attribute or tag corresponding to it. So in order to get the same effect of normal vs compact rendering of equations, you will have to pass the `display_value` argument explicitly to the OOML rendering request, as `to_omml(display_style: (true or false))`
```

Author: I will add a separate section for output as well.
